examples/efficientdet/efficientdet_building_blocks.py

```diff
@@ -0,0 +1,321 @@
+                config['conv_bn_act_pattern'],
+                config['separable_conv'],
+                config['act_type'],
+                weight_method=self.config['fpn_weight_method'],
+                data_format=self.config['data_format'],
+                name='fnode%d' % i)
+            self.fnodes.append(fnode)
+
+    def call(self, feats, training):
+
+        for fnode in self.fnodes:
+            feats = fnode(feats, training)
+        return feats
+
+
+def add_n(nodes):
+    """A customized add_n to add up a list of tensors."""
+    with tf.name_scope('add_n'):
+        new_node = nodes[0]
+        for n in nodes[1:]:
+            new_node = new_node + n
+        return new_node
+
+
+class FNode(tf.keras.layers.Layer):
+    """A keras layer implementing BiFPN node."""
+
+    def __init__(self,
+                 feat_level,
+                 inputs_offsets,
+                 fpn_num_filters,
+                 apply_bn_for_resampling,
+                 conv_after_downsample,
+                 conv_bn_act_pattern,
+                 separable_conv,
+                 act_type,
+                 weight_method,
+                 data_format,
+                 name='fnode'):
+        super().__init__(name=name)
+        self.feat_level = feat_level
+        self.inputs_offsets = inputs_offsets
+        self.fpn_num_filters = fpn_num_filters
+        self.apply_bn_for_resampling = apply_bn_for_resampling
+        self.conv_after_downsample = conv_after_downsample
+        self.conv_bn_act_pattern = conv_bn_act_pattern
+        self.separable_conv = separable_conv
+        self.act_type = act_type
+        self.weight_method = weight_method
+        self.data_format = data_format
+        self.resample_layers = []
+        self.vars = []
+
+    def fuse_features(self, nodes):
+        """Fuse features from different resolutions and return a weighted sum.
+
+        # Arguments
+            nodes: a list of tensorflow features at different levels.
+
+        # Returns
+            A tensor denoting the fused features.
+        """
+
+        dtype = nodes[0].dtype
+
+        if self.weight_method == 'attn':
+            edge_weights = [tf.cast(var, dtype=dtype) for var in self.vars]
+            normalized_weights = tf.nn.softmax(tf.stack(edge_weights))
+            nodes = tf.stack(nodes, axis=-1)
+            new_node = tf.reduce_sum(nodes * normalized_weights, -1)
+        elif self.weight_method == 'fastattn':
+            edge_weights = [tf.nn.relu(tf.cast(var, dtype=dtype))
+                            for var in self.vars]
+            weight_sum = add_n(edge_weights)
+            nodes = [nodes[i] * edge_weights[i] / (weight_sum + 0.0001)
+                     for i in range(len(nodes))]
+            new_node = add_n(nodes)
+        else:
+            raise ValueError('unknown weight_method %s' % self.weight_method)
+
+        return new_node
+
+    def _add_wsm(self, initializer, shape=None):
```

Comment:
what is wsm? be explicit with name functions.

Reply:
Yes, Octavio. It is actually weight scalar multidimensional. I changed it in the upcoming push to _add_scalar_mulitdimensional_weights